internal/BF_Base.ecl

```diff
@@ -0,0 +1,160 @@
+  * of expertly regularized BfTs.  Yet they require no regularization, work well
+  * with default parameters, and are as easy to use as RFs.
+  *
+  * BFs provide an early-stopping capability.  This allows the number of boosting
+  * iterations (i.e. maxLevels) to be specified at a high level (default 999),
+  * but only boosts for as many iterations as necessary to maximize accuracy.
+  * In normal practice, boosting will stop long before the default maxLevels is
+  * reached.
+  *
+  * <p>BF  provides myriad (i.e. multiple independent work-item) support,
+  * and support for both numeric and categorical independent data (i.e. X).
+  *
+  * @param X_in The independent Data in GenField format.
+  * @param Y_in The dependent Data in GenField format.
+  * @param maxLevels The maximum number of boosting iterations to perform.  This is
+  *         overridden by early stopping, and is primarily a failsafe in case the
+  *         data is non-separable.  Default (recommended) is 999.
+  * @param forestSize The number of trees to use in each Random Forest level.  The
+  *         default (recommended) is zero, which indicates that it should be
+  *         automatically determined by the software.
+  * @param maxTreeDepth The depth to which trees are grown.  Smaller numbers provide
+  *         weaker learners that are needed for BfT purposes.  The default (recommended)
+  *         is 20.
+  */
+EXPORT BF_Base(DATASET(GenField) X_in=DATASET([], GenField),
+                          DATASET(GenField) Y_in=DATASET([], GenField),
+                          UNSIGNED maxLevels=999,
+                          UNSIGNED forestSize=0,
+                          UNSIGNED maxTreeDepth=20
+                          ) := MODULE
+  SHARED clusterNodes := Thorlib.nodes();
+  // Forest size should normally be between 10 and 50.  If user specified 0,
+  // that indicates automatic choice.  If less than 10 nodes, we will use 10.
+  // If between 10 and 50 nodes, use the node count.  If > 50 nodes, limit it
+  // to 50.  In this way, we balance RF generalization with performance.
+  SHARED forestSizeAdj := IF(forestSize = 0, MIN(MAX(clusterNodes, 10), 50), forestSize);
+  // Features per node setting for RF.  If forestSize is 1, all features are used
+  // as there is no need to randomize the tree.  For larger forest sizes, we use 0
+  // to indicate that RF should automatically determine it.
+  SHARED fpn := IF(forestSizeAdj = 1, 999999, 0);
+
+  /**
+    * Convert the set of nodes describing the BfF to a Model Format
+    *
+    */
+  EXPORT DATASET(Layout_Model2) Nodes2Model(DATASET(BfTreeNodeDat) nodes) := FUNCTION
+    myRF := RF_Base();
+    empty_mod := DATASET([], Layout_Model2);
+    doOneRF(DATASET(Layout_Model2) model, UNSIGNED c) := FUNCTION
+      levNodes := nodes(bfLevel = c);
+      levMod := myRF.Nodes2Model(levNodes);
+      newModel := ModelOps2.Insert(model, levMod, [fModels]+[c]);
```

Comment:
You are appending the value of the counter as the last element in the array instead of using the set as a set.  Many ECL users will not be thinking of the set as an array.  I would expand the function comment to briefly describe the process of converting the BfF to model format.

Reply:
@johnholt 
Hey John, thanks for the review.  I'll add the comments you suggested on the next pull request.  It's good to hear from you!